python/lsst/pipe/base/all_dimensions_quantum_graph_builder.py

```diff
@@ -507,31 +508,39 @@ def from_builder(
                 yield result
 
     def log_failure(self, log: LsstLogAdapter) -> None:
-        """Emit a series of CRITICAL-level log message that attempts to explain
+        """Emit an ERROR-level log message that attempts to explain
         why the initial data ID query returned no rows.
 
         Parameters
         ----------
         log : `logging.Logger`
             The logger to use to emit log messages.
         """
-        log.critical("Initial data ID query returned no rows, so QuantumGraph will be empty.")
-        for message in self.common_data_ids.explain_no_results():
-            log.critical(message)
-        log.critical(
-            "To reproduce this query for debugging purposes, run "
-            "Registry.queryDataIds with these arguments:"
-        )
-        # We could just repr() the queryArgs dict to get something
-        # the user could make sense of, but it's friendlier to
-        # put these args in an easier-to-reconstruct equivalent form
-        # so they can read it more easily and copy and paste into
-        # a Python terminal.
-        log.critical("  dimensions=%s,", list(self.query_args["dimensions"].names))
-        log.critical("  dataId=%s,", dict(self.query_args["dataId"].required))
-        if self.query_args["where"]:
-            log.critical("  where=%s,", repr(self.query_args["where"]))
-        if "datasets" in self.query_args:
-            log.critical("  datasets=%s,", list(self.query_args["datasets"]))
-        if "collections" in self.query_args:
-            log.critical("  collections=%s,", list(self.query_args["collections"]))
+        # A single multiline log plays better with log aggregators like Loki.
+        buffer = StringIO()
+        try:
+            buffer.write("Initial data ID query returned no rows, so QuantumGraph will be empty.\n")
+            for message in self.common_data_ids.explain_no_results():
+                buffer.write(message)
+                buffer.write("\n")
+            buffer.write(
+                "To reproduce this query for debugging purposes, run "
+                "Registry.queryDataIds with these arguments:\n"
+            )
+            # We could just repr() the queryArgs dict to get something
+            # the user could make sense of, but it's friendlier to
+            # put these args in an easier-to-reconstruct equivalent form
+            # so they can read it more easily and copy and paste into
+            # a Python terminal.
+            buffer.write("  dimensions=%s," % list(self.query_args["dimensions"].names))
```

Comment:
We generally prefer `.format()` or f-strings in middleware. `ruff check --select UP031 .` triggers for this line. I haven't enabled UP031 in pipe_base yet because I didn't need to (this reminded me I planned to do that). I can fix this later when I turn on the check.

Reply:
I thought it would be better to keep the arguments analogous to what the direct logging calls would have been, but ok.